flink-yarn/src/main/java/org/apache/flink/yarn/YarnFlinkApplicationMasterRunner.java

```diff
@@ -0,0 +1,612 @@
+					"ApplicationMaster hostname variable %s not set", Environment.NM_HOST.key());
+			LOG.info("YARN assigned hostname for application master: {}", appMasterHostname);
+
+			// try to start the rpc service
+			// using the port range definition from the config.
+			final String amPortRange = config.getString(
+					ConfigConstants.YARN_APPLICATION_MASTER_PORT,
+					ConfigConstants.DEFAULT_YARN_JOB_MANAGER_PORT);
+
+			synchronized (lock) {
+				haServices = HighAvailabilityServicesUtils.createAvailableOrEmbeddedServices(config);
+				metricRegistry = new MetricRegistry(MetricRegistryConfiguration.fromConfiguration(config));
+
+				// ---- (2) init resource manager -------
+				resourceManagerRpcService = createRpcService(config, appMasterHostname, amPortRange);
+				resourceManager = createResourceManager(config);
+
+				// ---- (3) init job master parameters
+				jobMasterRpcService = createRpcService(config, appMasterHostname, amPortRange);
+				jobManagerServices = JobManagerServices.fromConfiguration(config, haServices);
+				jobManagerMetrics = new JobManagerMetricGroup(metricRegistry, jobMasterRpcService.getAddress());
+				jobMaster = createJobMaster(config);
+
+				// ---- (4) start the resource manager  and job master:
+				resourceManager.start();
+				LOG.debug("YARN Flink Resource Manager started");
+
+				// mark the job as running in the HA services
+				try {
+					haServices.getRunningJobsRegistry().setJobRunning(jobGraph.getJobID());
+				}
+				catch (Throwable t) {
+					throw new JobExecutionException(jobGraph.getJobID(),
+							"Could not register the job at the high-availability services", t);
+				}
+				jmLeaderElectionService.start(this);
+
+				// ---- (5) start the web monitor
+				// TODO: add web monitor
+			}
+			running = true;
+			while (running) {
+				Thread.sleep(100);
+			}
+			// everything started, we can wait until all is done or the process is killed
+			LOG.info("YARN Application Master finished");
+		}
+		catch (Throwable t) {
+			// make sure that everything whatever ends up in the log
+			LOG.error("YARN Application Master initialization failed", t);
+			shutdown(ApplicationStatus.FAILED, t.getMessage());
+			return INIT_ERROR_EXIT_CODE;
+		}
+
+		return 0;
+	}
+
+	// ------------------------------------------------------------------------
+	//  Utilities
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Validates a condition, throwing a RuntimeException if the condition is violated.
+	 * 
+	 * @param condition The condition.
+	 * @param message The message for the runtime exception, with format variables as defined by
+	 *				{@link String#format(String, Object...)}.
+	 * @param values The format arguments.
+	 */
+	private static void require(boolean condition, String message, Object... values) {
+		if (!condition) {
+			throw new RuntimeException(String.format(message, values));
+		}
+	}
+	protected RpcService createRpcService(
```

Comment:
The RpcServiceUtils.createRpcService has to use a specified port, can not use port range.